url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/LaneAgent.java

```diff
@@ -0,0 +1,320 @@
+package agent;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Semaphore;
+
+import DeviceGraphics.DeviceGraphics;
+import DeviceGraphics.PartGraphics;
+import GraphicsInterfaces.LaneGraphics;
+import agent.data.Part;
+import agent.interfaces.Lane;
+import factory.PartType;
+
+/**
+ * Lane delivers parts to the nest
+ * @author Arjun Bhargava
+ */
+public class LaneAgent extends Agent implements Lane {
+
+	public List<PartType> requestList = Collections
+			.synchronizedList(new ArrayList<PartType>());
+	public List<MyPart> currentParts = Collections
+			.synchronizedList(new ArrayList<MyPart>());
+
+	public PartType currentType;
+	
+	public int topLimit = 5;
+	public int lowerThreshold = 3;
+	public int extraRequestCount = 0;
+
+	public LaneStatus state;
+
+	String name;
+
+	public Semaphore animation = new Semaphore(0, true);
+
+	public class MyPart {
+		public Part part;
+		PartStatus status;
+
+		public MyPart(Part p) {
+			part = p;
+			status = PartStatus.BEGINNING_LANE;
+		}
+	}
+
+	public enum PartStatus {
+		BEGINNING_LANE, IN_LANE, END_LANE, TOLD_NEST, NEED_TO_DELIVER, DELIVERED
+	};
+
+	public enum LaneStatus {
+		FILLING, PURGING, WAITING, BROKEN, BROKEN_WHILE_PURGING;
+	};
+
+	FeederAgent feeder;
+	NestAgent nest;
+	LaneGraphics laneGUI;
+
+	public LaneAgent(String name) {
+		super();
+
+		this.name = name;
+		state = LaneStatus.FILLING;
+	}
+
+	@Override
+	public void msgINeedPart(PartType type) {
+		print("Received msgINeedPart");
+		requestList.add(type);
+		currentType = type;
+		stateChanged();
+	}
+
+	@Override
+	public void msgPurgeParts() {
+		print("Received msgPurgeParts");
+		if(state == LaneStatus.BROKEN_WHILE_PURGING || state == LaneStatus.BROKEN) {
+			state = LaneStatus.BROKEN_WHILE_PURGING;
+		} else {
+			state = LaneStatus.PURGING;
+		}
+		stateChanged();
+	}
+
+	@Override
+	public void msgHereIsPart(Part p) {
+		print("Received msgHereIsPart of type " + p.type.getName());
+		currentParts.add(new MyPart(p));
+		if (laneGUI != null) {
+			laneGUI.receivePart(p.partGraphics);
+		}
+
+		stateChanged();
+	}
+
+	@Override
+	public void msgGiveMePart() {
+		print("Received message give me part from nest");
+		synchronized (currentParts) {
+			for (MyPart p : currentParts) {
+				if (p.status == PartStatus.TOLD_NEST) {
+					p.status = PartStatus.NEED_TO_DELIVER;
+					break;
+				}
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 100:107, 115:122.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.